### PR TITLE
scvi-tools

### PR DESCRIPTION
**Pull request by tony-kuo**
Wrapper functions for scvi-tools workflows

### CODE REVIEW

File: pegasus/tools/scvitools.py [hunk 1]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
+    ) # register anndata
+    vae_ref = scvi.model.SCVI(adata, **arches_params) # obtain scvi model
+    vae_ref.train(max_epochs=max_epochs, use_gpu=use_gpu) # train model
+
+    data.obsm["X_scVI"] = vae_ref.get_latent_representation() # Get embedding
+    data.register_attr("X_scVI", "embedding") # Register X_scVI as an embedding
+
+    # semisupervised
+    if semisupervised_max_epochs is None:
+        semisupervised_max_epochs = np.min([round(np.sqrt(max_epochs)), 20])
```
Comment (bli25) [inline pegasus/tools/scvitools.py:317]: Is this default from scvitools code?

File: pegasus/tools/scvitools.py [hunk 2]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
+
+    if n_jobs == -1:
+        import multiprocessing
+        n_jobs = multiprocessing.cpu_count()
+    scvi.settings.num_threads = n_jobs # set n_jobs
+    scvi.settings.seed = random_state # set random_state, see [here](https://docs.scvi-tools.org/en/stable/_modules/scvi/_settings.html) for more details.
+
+    # unsupervised
+    if max_epochs is None:
+        max_epochs = np.min([round((20000 / len(adata.obs)) * 400), 400])
```
Comment (bli25) [inline pegasus/tools/scvitools.py:302]: I guess this code is from scvitools?

File: pegasus/tools/scvitools.py [hunk 3]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
+    adata = _gen_query_anndata(data, features, obs_columns, matkey) # gen AnnData
+
+    if n_jobs == -1:
+        import multiprocessing
+        n_jobs = multiprocessing.cpu_count()
+    scvi.settings.num_threads = n_jobs # set n_jobs
+    scvi.settings.seed = random_state # set random_state, see [here](https://docs.scvi-tools.org/en/stable/_modules/scvi/_settings.html) for more details.
+
+    if max_epochs is None:
+        max_epochs = np.min([round((20000 / len(adata.obs)) * 100), 100])
```
Comment (bli25) [inline pegasus/tools/scvitools.py:415]: How did we come with this default, is it from scvitools?

File: pegasus/tools/scvitools.py [hunk 4]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
+    continuous_covariate_keys: ``List[str]``
+        A list of obs keys listing continuous covariates that should be corrected for, default is ``None``.
+    use_gpu: ``str | int | bool | None``
+        Use default GPU if available (if None or True), or index of GPU to use (if int), or name of GPU (if str, e.g., ‘cuda:0’), or use CPU (if False).
+
+    Returns
+    -------
+    Update ``data.obsm``:
+        * ``data.obsm['X_scanVI']``: The embedding calculated by scanVI.
+        * ``data.obsm[predictions]``: The predicted labels by scanVI.
```
Comment (bli25) [inline pegasus/tools/scvitools.py:385]: Do we need single quotes around `predictions`?

File: setup.py [hunk 5]
```diff
@@ -55,7 +55,8 @@
         forceatlas=["forceatlas2-python"],
         mkl=["mkl"],
         rpy2=["rpy2"],
-        all=["fitsne", "louvain", "scanorama", "torch", "harmony-pytorch", "nmf-torch", "rpy2", "forceatlas2-python"]
+        scvi=["scvi-tools"],
+        all=["fitsne", "louvain", "scanorama", "torch", "harmony-pytorch", "nmf-torch", "rpy2", "forceatlas2-python", "scvi-tools"]
     ),
     python_requires="~=3.6",
```
Comment (bli25) [inline setup.py]: Please change it to 3.7.

**Review by bli25 — CHANGES_REQUESTED** [review]

File: pegasus/tools/scvitools.py (hunk 2 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py:302]: Yes this is the default from scvi-tools

File: pegasus/tools/scvitools.py (hunk 3 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py:415]: This was not the default from scvi-tools.  This was just something I used as a placeholder based on the tutorial using 100 epochs for this step

File: pegasus/tools/scvitools.py (hunk 1 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py:317]: This was not the default from scvi-tools. This was just something I used as a placeholder based on the tutorial using 20 epochs for this step

File: pegasus/tools/scvitools.py (hunk 4 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py:385]: predictions is a variable, the default of which is "predictions".  Would you prefer a fixed obsm key name?

File: pegasus/tools/scvitools.py (hunk 3 above)
Comment (bli25) [reply pegasus/tools/scvitools.py:415]: Sounds good.

File: pegasus/tools/scvitools.py (hunk 4 above)
Comment (bli25) [reply pegasus/tools/scvitools.py:385]: What you have is good.

File: pegasus/tools/scvitools.py (hunk 1 above)
Comment (bli25) [reply pegasus/tools/scvitools.py:317]: Fine.

File: setup.py (hunk 5 above)
Comment (tony-kuo) [reply setup.py]: done, in the latest commit

File: pegasus/tools/scvitools.py [hunk 6]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
+        obs_columns.append(batch)
+    if categorical_covariate_keys is not None and categorical_covariate_keys:
+        obs_columns.extend(categorical_covariate_keys)
+    if continuous_covariate_keys is not None and continuous_covariate_keys:
+        obs_columns.extend(continuous_covariate_keys)
+
+    adata = _gen_anndata(data, features, obs_columns, matkey) # gen AnnData
+
+    if n_jobs == -1:
+        import multiprocessing
```
Comment (bli25) [inline pegasus/tools/scvitools.py]: Please use pg.utils.eff_n_jobs (see `nmf.py` as an example), which set physical cores when possible.

File: pegasus/tools/scvitools.py [hunk 7]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
```
Comment (bli25) [inline pegasus/tools/scvitools.py]: Please use pg.utils.eff_n_jobs (see `nmf.py` as an example), which set physical cores when possible.

File: pegasus/tools/scvitools.py [hunk 8]
```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
...
+    if categorical_covariate_keys is not None and categorical_covariate_keys:
+        obs_columns.extend(categorical_covariate_keys)
+    if continuous_covariate_keys is not None and continuous_covariate_keys:
+        obs_columns.extend(continuous_covariate_keys)
+
+    features = scvi.model.SCANVI.prepare_query_anndata(None, dir_path, return_reference_var_names=True)
+    adata = _gen_query_anndata(data, features, obs_columns, matkey) # gen AnnData
+
+    if n_jobs == -1:
+        import multiprocessing
```
Comment (bli25) [inline pegasus/tools/scvitools.py]: Please use pg.utils.eff_n_jobs (see `nmf.py` as an example), which set physical cores when possible.

**Review by bli25 — CHANGES_REQUESTED** [review]

File: pegasus/tools/scvitools.py (hunk 6 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py]: done in new commit

File: pegasus/tools/scvitools.py (hunk 7 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py]: done in new commit

File: pegasus/tools/scvitools.py (hunk 8 above)
Comment (tony-kuo) [reply pegasus/tools/scvitools.py]: done in new commit

**Review by bli25 — APPROVED** [review]

**Review by bli25 — APPROVED** [review]

**Pull request merged**